Artisan generators

jstoone opened the pull request:
This is the initial work for what's been discussed in #25.

---

## The good ol' todo
- [x] Create Generators
  - [x] Hydrator
  - [x] Resource
  - [x] Schema
  - [x] Search *(Eloquent only)*
  - [x] Validators
  - [x] Request
  - [x] Resource *(Creates all of the above)*
- [x] Create eloquent implementation stubs
  - [x] `stubs/eloquent/hydrator.stub`
  - [x] `stubs/eloquent/schema.stub`
  - [x] `stubs/eloquent/search.stub` *(Eloquent only)*
- [x] Create non-eloquent implementation stubs
  - [x] `stubs/abstract/hydrator.stub`
  - [x] `stubs/abstract/schema.stub`
- [x] Create implementation agnostic stubs
  - [x] `stubs/independent/validators.stub`
  - [x] `stubs/independent/equest.stub`
- [x] Make it possible to not use eloquent
- [x] Make it possible to change namespace format as mentioned in #25
- [x] Add extra flags to `ResourceMakeCommand`
  - [x] Implement `--only=foo,bar`
  - [x] Implement `--except=foo,bar`
- [x] Only run `SearchMakeCommand` if `eloquent == true`
- [x] Write documentation

lindyhopchris commented on config/json-api.php:
```diff
@@ -143,4 +143,37 @@
+    */
+    'generator' => [
+        'namespace' => 'JsonApi',
+        'by_resource' => true,
```
Can these be dasherized rather than underscored? I.e. `by-resource`. Just keeps it consistent with how the rest of the config file is written.

lindyhopchris commented on src/Console/Commands/JsonApiGeneratorCommand.php:
```diff
@@ -0,0 +1,176 @@
+    protected function getOptions()
+    {
+        return [
+            ['eloquent', 'e', InputOption::VALUE_OPTIONAL, 'Use eloquent as adapter.'],
```
For the eloquent option, I think have `--eloquent` and `--no-eloquent` as boolean options. If `--eloquent` and `--no-eloquent` both return `false` (i.e. neither have been provided) then fall back to the value in config. I.e...

``` php
protected function isEloquent()
{
	if ($this->isIndependent) {
		return false;
	}
	
	if ($this->option('no-eloquent')) {
		return false;
	}
	
	return $this->option('eloquent') ?: $this->useEloquent;
}
```

lindyhopchris commented on src/Console/Commands/JsonApiGeneratorCommand.php:
```diff
@@ -0,0 +1,176 @@
+     *
+     * @var mixed
+     */
+    protected $isIndependant = false;
```
Typo in variable name - should be `isIndependent`

lindyhopchris commented on src/Console/Commands/JsonApiGeneratorCommand.php:
```diff
@@ -0,0 +1,176 @@
+    protected function getStub()
+    {
+        if($this->isIndependant) {
+            return __DIR__.'/stubs/'.strtolower($this->type).'.stub';
```
Can we move the `stubs` directory to the root project folder. I have a strong belief that the `src` folder should just be for classes!

lindyhopchris commented on src/Console/Commands/stubs/eloquent/request.stub:
```diff
@@ -0,0 +1,24 @@
+    protected $hasMany = [
+        //
+    ];
+}
```
For the Eloquent request stub, can we add the following:

``` php
    /**
     * @var array
     */
    protected $allowedFilteringParameters = [
        'id',
    ];
```

Because by default the search supports that as a filtering parameter.

lindyhopchris commented:
PS: haven't used this review feature before! I do appreciate you're sharing unfinished work!!

jstoone commented:
Thanks a bunch for giving good feedback on my code. I think you did very well on using the feature. :)

I'll keep you posted as I do progress.

jstoone commented:
Changes from the latest commit fixes all of your comments from the review:
- [x] Use dashes in config
- [x] Fix typo in `$isIndependent` variable
- [x] Move stubs to root folder
- [x] Remove `--eloquent` option from independent generators
- [x] Clean up `isEloquent()` logic
- [x] Add `getResourceType()`-method to schema stub
- [x] Add `allowedFilterParameters()` to eloquent stub

Extra changes
- [x] Cleaned up `getStub()`-method
- [x] Request is now independent
- [x] Simplified command registration in ServiceProvider
- [x] `independent` is now a type, and has its own folder.

jstoone commented:
Alright! The generators are now feature complete, and ready for you to lay your eyes upon them. 🎃 

Change log since last commit-burst:
- [x] `--no-eloquent`-flag now works
- [x] You can now how choose between two namespace models:
  - [x] By resource name e.g: `JsonApi\Tasks\Schema`
  - [x] By class type e.g: `JsonApi\Schemas\Task`
- [x] The `Search` class now only gets generated if configured for eloquent
- [x] A bit of documentation added
- [x] Found a config item that still was using an underscore
  - [x] `'use_eloquent` => 'use-eloquent'`

lindyhopchris commented on config/json-api.php:
```diff
@@ -3,10 +3,7 @@
-    | JSON API Namespaces
-    |--------------------------------------------------------------------------
-    |
+    |-------------------------------------------------------------------------- | JSON API Namespaces |-------------------------------------------------------------------------- |
```
Can this heading be put back to the format that it was before?

lindyhopchris commented on config/json-api.php:
```diff
@@ -143,4 +140,37 @@
+    |--------------------------------------------------------------------------
+    |
+    | With this package comes a set of handy generators. Which makes it possible
+    | to easily generate every class needed to implement an api resource.
```
Can this sentence be changed to:

```
This package supplies a set of handy generators. These make it possible
to easily generate every class needed to implement a JSON API resource.
```

lindyhopchris commented on config/json-api.php:
```diff
@@ -143,4 +140,37 @@
+    | we have included a few handy configuration options, so that the generators
+    | can follow your workflow.
+    |
+    | `namespace`       = The folder in which you will be storing everyting
```
typo: `everything`

lindyhopchris commented on config/json-api.php:
```diff
@@ -143,4 +140,37 @@
+    |                           e.g. \App\JsonApi\Schemas\{User, Post, Comment}
+    |                           e.g. \App\JsonApi\Requests\{User, Post, Comment}
+    |
+    | `use_eloquent`    = Whether you are using Eloquent ORM in this app.
```
Suggested re-wording for the explanation here:

```
Whether your JSON API resources relate to Eloquent models or not. 
You can override the setting here when running a generator. If the 
setting here is `true` running a generator with `--no-eloquent` will 
override it; if the setting is `false`, then `--eloquent` is the override.
```

jstoone replied:
Don't know why this has not been marked as complete, but I've changed the text as mentioned. I hope I didn't misunderstand you. :)

lindyhopchris commented on config/json-api.php:
```diff
@@ -143,4 +140,37 @@
+        'namespace' => 'JsonApi',
+        'by-resource' => true,
+        'use-eloquent' => true,
+    ]
```
requires a trailing comma for consistency with coding style.

lindyhopchris commented on src/Console/Commands/JsonApiGeneratorCommand.php:
```diff
@@ -0,0 +1,250 @@
+    /**
+     * Get the stub for specific generator type
+     *
+     * @param string mentationType
```
Typo in doc block variable name

lindyhopchris commented on src/Console/Commands/JsonApiGeneratorCommand.php:
```diff
@@ -0,0 +1,250 @@
+            '/',
+            $implementationType,
+            '/',
+            $this->type,
```
This needs to be `lcfirst($this->type)` otherwise it fails to find the file on systems that have case-sensitive file names

jstoone replied:
I'm still amazed you caught this, flew right past me that `$this->type` of course is manually `ucfirst`'ed.

lindyhopchris commented on src/Console/Commands/JsonApiGeneratorCommand.php:
```diff
@@ -0,0 +1,250 @@
+    protected function replaceResourceType(&$stub, $resource)
+    {
+        $stub = str_replace(
+            'dummyResourceType', camel_case($resource), $stub
```
Change to: `snake_case($resource, '-')`

Because if generating a multi-word resource type, the JSON API convention is for it to be a dasherized string

lindyhopchris commented on src/Console/Commands/JsonApiGeneratorCommand.php:
```diff
@@ -0,0 +1,250 @@
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Input\InputArgument;
+
+abstract class JsonApiGeneratorCommand extends LaravelGeneratorCommand
```
Sorry for not spotting this before, but the convention in the package would be to call this class `AbstractGeneratorCommand`.

lindyhopchris commented on stubs/abstract/hydrator.stub:
```diff
@@ -0,0 +1,28 @@
+use CloudCreativity\JsonApi\Hydrator\RelatedHydratorTrait;
+use CloudCreativity\JsonApi\Contracts\Hydrator\HydratesRelatedInterface;
+
+class Hydrator extends AbstractHydrator implements HydratesRelatedInterface
```
This stub should be as follows:

``` php
<?php

namespace DummyNamespace;

use CloudCreativity\JsonApi\Contracts\Hydrator\HydratesRelatedInterface;
use CloudCreativity\JsonApi\Contracts\Object\StandardObjectInterface;
use CloudCreativity\JsonApi\Hydrator\AbstractHydrator;
use CloudCreativity\JsonApi\Hydrator\RelatedHydratorTrait;

class Hydrator extends AbstractHydrator implements HydratesRelatedInterface
{

    use RelatedHydratorTrait;

    /**
     * @param StandardObjectInterface $attributes
     * @param $record
     * @return void
     */
    protected function hydrateAttributes(StandardObjectInterface $attributes, $record)
    {
        // TODO: Implement hydrateAttributes() method.
    }

}
```

lindyhopchris commented on stubs/abstract/schema.stub:
```diff
@@ -0,0 +1,30 @@
+    /**
+     * @var array
+     */
+    protected $attributes = [
```
Remove this property as that's to do with the `EloquentSchema`

lindyhopchris commented on stubs/abstract/schema.stub:
```diff
@@ -0,0 +1,30 @@
+    {
+        return self::RESOURCE_TYPE;
+    }
+}
```
The following methods can be added to this stub:

``` php

    /**
     * @param object $resource
     * @return mixed
     */
    public function getId($resource)
    {
        // TODO
    }

    /**
     * @param object $resource
     * @return mixed
     */
    public function getAttributes($resource)
    {
        // TODO
    }
```

lindyhopchris commented on stubs/eloquent/search.stub:
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace DummyNamespace
```
Missing semi-colon

lindyhopchris commented on stubs/independent/validators.stub:
```diff
@@ -0,0 +1,43 @@
+<?php
+
+namespace DummyNamespace
```
Missing semi-colon

lindyhopchris commented on src/Console/Commands/HydratorMakeCommand.php:
```diff
@@ -0,0 +1,27 @@
+     *
+     * @var string
+     */
+    protected $description = 'Create a new json-api resource hydrator';
```
On all of the descriptions, can we change `json-api` to `JSON API`

lindyhopchris commented on src/ServiceProvider.php:
```diff
@@ -9,8 +9,7 @@
  *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
```
Can you put this back to what it was?

jstoone commented:
### Awesome review, awesome job!
I've now run through every review you've made and change-log is below. You are a very sweet reviewer, and have a very sharp eye! 

Lucky me that you've almost written down, to the word, what changes I should make. Easy peasy. Such a pleasure working like this. 👍 

### Changelog
- [x] Corrected `config/json-api.php`
    - [x] Sentences/re-wording
    - [x] Spelling
    - [x] Formatting
- [x] Renamed `JsonApiGeneratorCommand` to `AbstractGeneratorCommand`
- [x] Change resource type from using `camel_case` to `snake_case`
- [x] General typos
- [x] Lowercase the generator type when creating stub path
- [x] Correct abstract hydrator stub
- [x] Correct abstract schema stub
- [x] Add missing semi-colon to namespace in:
    - [x] Independent validators stub
    - [x] Eloquent Search stub
- [x] Change generator descriptions to "JSON API"
- [x] Reformat licence in in ServiceProvider

lindyhopchris commented:
@jstoone awesome work - thanks so much for doing this, it's a fantastic addition to the package and something that I wouldn't have got to for a while because of work deadlines.

If you have any other proposal for additions to the package do let me know via issues! (Not saying you'd have to do the pull requests but would be great to hear your ideas!)

jstoone commented:
@lindyhopchris you are very welcome! As I said before, it was a pleasure doing this and working with you. Let's see what we'll figure out next. :)
